### DISCUSSION TITLE
Rich - under the hood

### DISCUSSION THREAD

**Discussion by FedeClaudi** [Textualize/rich#1937] [Q&A] [answered]
Hey Will, 

I've been looking at Rich's code more closely. trying to get an idea how does it *actually* work "under the hood" - and I'm not sure I got it right, could you please clarify some doubts?

Looking at `Console.print`, it [extends](https://github.com/Textualize/rich/blob/972dedff546a60642db36831dbb8aaab19077a2c/rich/console.py#L1635) `buffer` with `Segment`s. The `Console._thread_locals` buffer is an instance of `ConsoleThreadLocal` which stores a list of of `Segment`s. Next, `Console._render_buffer` returns a `str` with the rendered `Segments` that are in the buffer. Given that `Console.print` uses `with self`, when the context is closed it calls `_check_buffer` and that returns the string to be printed. 

Does this sound about right, or am I missing something?  Is then the 'magic' in making fancy stuff like `Table`s and `Panel`s  about creating the appropriate `Segment`s?

Thanks for your time!

**Comment by willmcgugan** [comment, marked as answer]
Hi Federico,

Your assessment is accurate. Ultimately all the renderables will generate a sequence of Segments.

The advantage of Segments is that they can be further processed. For instance the table can render something and process the segments to fit within the cells. It also makes it possible to target the output to either the console or HTML. It would be very difficult to do any of that if you generate text with embedded ansi escape codes.

Will

**Reply by FedeClaudi** [reply]
Thank you!

**Comment by FedeClaudi** [comment]
Hi Will,

I've recently started implementing some of rich's features in a julia library: https://github.com/FedeClaudi/Term.jl. 
I don't see this ever getting anywhere near the level of `rich`, I'm doing it mostly for fun/my own use. But since I'm using a lot of the work that you've put into `rich` to speed develop this library I just wanted to check that you're okay with that. Of course there will always be a reference to rich and credit to you for developing `rich`.

**Reply by willmcgugan** [reply]
Hi Federico,

I am very okay with that. There are already [ports](https://spectreconsole.net/) to other languages, and I think its terrific to bring Rich or something Rich-like to other languages.

As always, let me know if you have any questions re the implementation!

**Reply by FedeClaudi** [reply]
Awesome, I'll bug you if I get stuck with something! If there's anything that you think would be particularly hard to figure out feel free to give any pointers!

**Reply by willmcgugan** [reply]
Textual.render may need a second glance. Implementation is not terribly obvious.

**Comment by FedeClaudi** [comment]
HI will, 

I hope all is well. I realize I'm asking, for your time to answer questions that are not strictly Rich related - so if you're busy please just ignore the following. 

As you know, I'm developing [Term](https://github.com/FedeClaudi/Term.jl) in Julia which implements some of `rich`'s features. 

Lately I've been re-working the progress bar functionality, and now the progress bars sit at the bottom of the terminal with any `stdout` output being displayed about. This means that, in addition to printing a whole lot of text to the terminal, I'm also using ANSI commands code to change scrollregions, move the cursor etc... 
Which leads me to the question of: **how the heck do I write tests to see if it is all running correctly**? I'm guessing you've solved similar issues in rich, hence my question. 

I can see a few options:
1. just check that it all runs without errors: it doesn't guarantee that the visual display is correct
2. test each individual component, and hope it all works correctly together
3. save a text file with the sequence of stdout output and check that they match.

The last option is the most accurate, but it requires a bunch of work any time something is changed. So it doesn't test for bug just for changes. 

Any advice?

![Uploading ezgif.com-gif-maker(9).gif…]()

**Reply by willmcgugan** [reply]
A combination of 2 and 3 I'd say.

You definitely want to test each component by checking the state matches expectations.

But ultimately you will need to render it out to ansi codes, and check the output matches the expected render. You are absolutely right about the downside that you will have to update things when something changed, because there is no way to assert something looks correct -- only that it has changed. You may also have to manage the fact that the output may be different across platforms / environment variables.

**Reply by FedeClaudi** [reply]
Gotcha, thank you!
